src/smt/smt_engine.cpp

```diff
@@ -1208,11 +1212,45 @@ void SmtEngine::setDefaults() {
   {
     d_logic = LogicInfo("QF_NIA");
   }
-  else if ((d_logic.getLogicString() == "QF_UFBV"
-            || d_logic.getLogicString() == "QF_ABV")
-           && options::bitblastMode() == theory::bv::BITBLAST_MODE_EAGER)
+
+  // set options about ackermannization
+  if (options::ackermann())
   {
-    d_logic = LogicInfo("QF_BV");
+    if (options::produceModels()
+        && (d_logic.isTheoryEnabled(THEORY_ARRAYS)
+            || d_logic.isTheoryEnabled(THEORY_UF)))
+    {
+      if (options::produceModels.wasSetByUser())
+      {
+        throw OptionException(std::string(
+            "Ackermannization currently does not support model generation."));
+      }
+      Notice() << "SmtEngine: turn off ackermannization to support model"
+               << "generation" << endl;
+      options::ackermann.set(false);
```

Comment:
I think that it is slightly better to disable model generation in that case because it makes the option checking easier (otherwise we are checking whether incremental solving has been enabled below even though we just disabled ackermannization).

Reply:
The option of produceModel is read-only. Seems I cannot set it to false.

Reply:
Ok, let's do it the other way around in that case. Let's have a separate block (i.e. outside of the current `if (options::ackermann())` block) that does something like:
```
if (options::produceModels()) {
  if (options::ackermann() && (d_logic.isTheoryEnabled(THEORY_ARRAYS) || d_logic.isTheoryEnabled(THEORY_UF)) {
    if (options::ackermann.wasSetByUser()) {
      // throw error
    }
    // disable ackermannizations
  }
}
```

This is cleaner (IMHO :) ) than rechecking below whether ackermannization is enabled.